### DISCUSSION TITLE
Performance API

### DISCUSSION THREAD

**Discussion by philipphofmann** [getsentry/sentry-cocoa#936] [General]
Discussion about the performance API here https://github.com/getsentry/sentry-cocoa/pull/909

**Comment by philipphofmann** [comment]
I would propose a few changes then to the whole API. For simplicity, I skipped `Sentry` in the naming.

1. `Span` shouldn't inherit from `SpanContext`, but instead, have a property to it. 
2. Add protocols for `Span` and `Transaction` with NoOps
3. Rename the classes `Span` and `Transaction` to something like `InternalSpan` and `InternalTransaction`
4. Extract a base class `BaseEvent` from `Event` and let both `Transaction` and `Event` inherit from it.

This would look something like this.
![Performance API](https://user-images.githubusercontent.com/2443292/107020867-ab86cb00-67a3-11eb-9fb3-d0314bcd70a2.png)

**Reply by marandaneto** [reply]
1) the reason is serialization over reflection (Gson-Java), if you are in control of your serialization, which is the case in Cocoa, that makes total sense.
ps: Custom Gson adapters would be a solution.
2) and 4 are a must IMO.
3) I don't see any benefit tbh, why would we do that?

**Reply by philipphofmann** [reply]
3. Because we don't want to prefix the protocols/interfaces with `I` because that's not common in Objective-C. Do you agree @brustolin?

**Reply by brustolin** [reply]
Yes, the "I" prefix is not a pattern in objc, we don't want that. But I didn't understand the need to create the "internal" and "NoOp" versions of the class.

If Span is no longer subclass of SpanContext, the hidden initializers problem is solved.

**Reply by bruno-garcia** [reply]
They are not going to be Internal necessarily are those?

**Reply by marandaneto** [reply]
@philipphofmann but we talked that in objC an interface and a class may have the same naming, so we don't need to add `Internal`

**Reply by marandaneto** [reply]
@brustolin we need a NoOp version of Transaction and Span because we don't want our users doing this:

```
val transaction = startTransaction(..)
if (transaction != null) {
  // do something
}
```

the idea is like the `SentryId`
https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentrySDK.m#L97-L100

even if something went wrong, return a non null instance of SentryId, but zeroed https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryId.m#L6

so its not a valid sentryId, but its not null, so its just a NoOp.
the same would be for Transaction/Span.

**Reply by bruno-garcia** [reply]
The only problem is that we can't propagate this id on outgoing http requests. With the "sampled=false" part, so downstream services could be creating a completely new transaction without a parent.
The approach on python was to create the id regardless, with sampled false. That can be used to link errors at least (even though transaciton was unsampled)

**Reply by marandaneto** [reply]
also fine with creating a Transaction with sampled=false only (instead of NoOp-for now).
I still prefer a NoOp, so we save some allocation, its a performance SDK, so every bit matters IMO.

**Reply by marandaneto** [reply]
> That can be used to link errors at least (even though transaciton was unsampled)

that's the point, if the transaction is unsampled, you link errors but transactions wont be sent anyway, so what's the benefit I am missing?

**Reply by philipphofmann** [reply]
> we need a NoOp version of Transaction and Span because we don't want our users doing this
> val transaction = startTransaction(..)
> if (transaction != null) {
>   // do something
> }

@marandaneto why would `startTransaction(..)` return `null`? I would be happy to skip the `NoOpTransaction`.

**Reply by marandaneto** [reply]
@philipphofmann if the SDK is eg disabled

**Reply by philipphofmann** [reply]
That makes sense, but do you think it's worth the effort of adding an interface and a NoOP implementation just for the case when the SDK is disabled? Yes, a NoOp saves the user a lot of extra work, but we can discard the transactions when calling the hub. What do you think @bruno-garcia and @brustolin ?

**Reply by marandaneto** [reply]
@philipphofmann that's what .NET did, the downside is, allocation, we are allocating objects that are not gonna be used, it'd be cheaper to have a NoOp that does nothing.

**Reply by philipphofmann** [reply]
I think a more straightforward API is more important than optimizing for when the SDK is disabled. I think it's okay that the SDK still allocates a few objects in that case. The most important thing is that the actual transaction is fast IMO.

**Reply by bruno-garcia** [reply]
> > That can be used to link errors at least (even though transaciton was unsampled)
> 
> that's the point, if the transaction is unsampled, you link errors but transactions wont be sent anyway, so what's the benefit I am missing?

Benefit is linked errors front-end and back-end

**Reply by bruno-garcia** [reply]
> @philipphofmann that's what .NET did, the downside is, allocation, we are allocating objects that are not gonna be used, it'd be cheaper to have a NoOp that does nothing.

Saving on allocations isn't the only reason to use a NoOp, it's also the "null pattern" which was also nice when languages didn't have null annotations and users could end up with NPE if they weren't careful.
I often point out to saved allocation when I spot one but doesn't mean we need to drive the design solely by that mindset, I honestly am thorn about NoOp transactions because of the limitation of not propagating the trace-id and limiting the ability to link errors in different services. The allocation of a single object is trivial, specially compared with the value we get from linking errors.

I'm OK with keeping our options open here and always return a transaction instance, with `sampled=false` but an id. Optimization could exist in avoid creating Spans when the transactions is unsampled? There I see no value in allocating since the spans won't be captured. But a `trace-id` and propagation we should aim getting to work.

**Comment by philipphofmann** [comment]
According to @mitsuhiko, we should try to keep transactions private and only expose spans if possible.

**Comment by philipphofmann** [comment]
Why is the description for `startChild` nullable when we have an overload without the description? Can we make it non-nullable?

```swift
span.startChild(operation: "Child 1 Child", description: nil)
transaction.startChild(operation: "Child 2", description: nil)
```

**Reply by philipphofmann** [reply]
Is addressed in https://github.com/getsentry/sentry-cocoa/pull/1059

**Comment by armcknight** [comment]
This has been stale for a while so closing it as outdated.
